apiserver/facades/client/resources/repository.go

```diff
@@ -43,58 +43,183 @@ type CharmHub interface {
+	if !ok {
+		// This indicates that AddPendingResources() was called for
+		// a resource the charm backend doesn't know about (for the
+		// relevant charm revision).
+		return res, nil
+	}
+
+	if res.Revision < 0 {
+		// The caller wants to use the charm backend info.
+		return storeRes, nil
+	}
+	if res.Revision == storeRes.Revision {
+		// We don't worry about if they otherwise match. Only the
+		// revision is significant here. So we use the info from the
+		// charm backend since it is authoritative.
+		return storeRes, nil
+	}
+	if res.Fingerprint.IsZero() {
+		// The caller wants resource info from the charm backend, but with
+		// a different resource revision than the one associated with
+		// the charm in the backend.
+		return c.client.ResourceInfo(id.URL, id.Origin, res.Name, res.Revision)
+	}
+	// The caller fully-specified a resource with a different resource
+	// revision than the one associated with the charm in the backend. So
+	// we use the provided info as-is.
+	return res, nil
+}
+
 // TODO hml 2020-12-2
 // charmHubClient needs to be "caching" like the charmstoreclient is??
 type charmHubClient struct {
+	resourceClient
 	client CharmHub
 	id     CharmID
 }
 
-// ResolveResources, looks at the provided, charmhub and backend (already downloaded)
-// resources to determine which to use.  Provided (uploaded) take precedence.  If
-// charmhub has a newer resource than the back end, use that.
-// TODO: (hml) 2020-12-03
-// this logic will need refinement as work to incorporate charmhub resources continues.
-// Right now, just take the uploaded resources and add charmhub resources where resource
-// is missing
-func (ch *charmHubClient) ResolveResources(uploadedResources []charmresource.Resource) ([]charmresource.Resource, error) {
-	chResources, err := ch.listResources()
+func newCharmHubClient(client CharmHub, id CharmID) *charmHubClient {
+	c := &charmHubClient{
+		client: client,
+		id:     id,
+	}
+	c.resourceClient = resourceClient{
+		client: c,
+	}
+	return c
+}
+
+// ResolveResources, looks at the provided, charmhub and backend (already
+// downloaded) resources to determine which to use. Provided (uploaded) take
+// precedence. If charmhub has a newer resource than the back end, use that.
+func (ch *charmHubClient) ResolveResources(resources []charmresource.Resource) ([]charmresource.Resource, error) {
+	storeResources, err := ch.listResources()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	for _, res := range uploadedResources {
-		chResources[res.Name] = res
+	resolved, err := ch.resolveResources(ch.id, resources, storeResources)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	results := make([]charmresource.Resource, len(chResources))
-	var i int
-	for _, res := range chResources {
-		results[i] = res
-		i += 1
+	return resolved, nil
+}
+
+func (ch *charmHubClient) ResourceInfo(curl *charm.URL, origin corecharm.Origin, name string, revision int) (charmresource.Resource, error) {
```

Comment:
Both the charmHubClient and charmStoreClient have id embedded in them, and ResourceInfo takes an origin as input.  Suggest ResourceInfo doesn't take a origin as input, or that clients do not need the id.

Reply:
I've moved stuff around, I'd rather not pass in the CharmID and be more explicit about URL and Origin here. Only because Metadata is hanging off CharmID and people would assume that it has an important role when it comes to locating a resource via `ResourceInfo`, which is not the case.

Reply:
Quick offline conversation.  We'll use the origin passed in consistently, and not the one embedded in the resourceClient.